### PR TITLE
Make commit multithreaded

### PR DESCRIPTION
**Pull request by yarkivaev**

### CODE REVIEW

File: lib/src/ssyp_db.cpp [hunk 1]
```diff
@@ -70,14 +75,50 @@ class SsypDB : public ISsypDB {
...
+    bool stopThread = false;
+    bool isCommitThread = false;
+    std::thread commitThread;
+    std::mutex mutex;
+
+    void InitializeCommitThread() {
+        if (isCommitThread) {
+            return;
+        }
+        commitThread = std::thread(&SsypDB::CommitThreadCycle, this);
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: У нас сейчас есть вполне определенный класс в котором мы работаем. Так что лучше просто проинициализироваться в конструкторе. И не делать detach, а сделать join в деструкторе (так как наша функция использует внутренности класса, это уже становится довольно опасно, класс может быть деинициализирован в тот момент, когда мы здесь пытаемся что-то закоммитить)

File: lib/src/ssyp_db.cpp [hunk 2]
```diff
@@ -70,14 +75,50 @@ class SsypDB : public ISsypDB {
...
+        commitThread.detach();
+    }
+    void CommitThreadCycle() {
+        while (!stopThread) {
+            if (!transactionPtrQueue_.empty()) {
+                std::vector<std::pair<ITransactionPtr, std::promise<CommitStatus>> > localTransactionQueue;
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: Тут можно написать попроще
```c++
decltype(transactionPtrQueue_) localTransactionQueue
```

File: lib/src/ssyp_db.cpp [hunk 3]
```diff
@@ -70,14 +75,50 @@ class SsypDB : public ISsypDB {
...
+                {
+                    std::lock_guard<std::mutex> lock(mutex);
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: `<std::mutex>` писать не обязательно

File: lib/src/ssyp_db.cpp [hunk 4]
```diff
@@ -70,14 +75,50 @@ class SsypDB : public ISsypDB {
...
+                    std::swap(localTransactionQueue, transactionPtrQueue_);
+                }
+
+                for (auto it = localTransactionQueue.begin(); it != localTransactionQueue.end(); it++) {
+                    (it->second).set_value(datamodel_->Commit(((Transaction*)((it->first).get()))->ops)
+                        ? CommitStatus::Success
+                        : CommitStatus::Error);
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: Не пиши огромных однострочников, плз

File: lib/src/ssyp_db.cpp [hunk 5]
```diff
@@ -70,14 +75,50 @@ class SsypDB : public ISsypDB {
...
+    void CommitThreadCycle() {
+        while (!stopThread) {
+            if (!transactionPtrQueue_.empty()) {
+                std::vector<std::pair<ITransactionPtr, std::promise<CommitStatus>> > localTransactionQueue;
+                {
+                    std::lock_guard<std::mutex> lock(mutex);
+                    std::swap(localTransactionQueue, transactionPtrQueue_);
+                }
+
+                for (auto it = localTransactionQueue.begin(); it != localTransactionQueue.end(); it++) {
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]:
```c++
for (auto& [tx, promise] : localTransactionQueue) { ... }
```

File: lib/src/ssyp_db.cpp [hunk 6]
```diff
@@ -70,14 +75,50 @@ class SsypDB : public ISsypDB {
...
+                    (it->second).set_value(datamodel_->Commit(((Transaction*)((it->first).get()))->ops)
+                        ? CommitStatus::Success
+                        : CommitStatus::Error);
+                }
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(5));
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: Попробуй все-таки сделать через `std::condition_variable`

File: lib/src/ssyp_db.cpp [hunk 7]
```diff
@@ -70,14 +75,50 @@ class SsypDB : public ISsypDB {
...
+        }
+        std::lock_guard<std::mutex> lock(mutex);
+        transactionPtrQueue_.push_back(std::make_pair(tx, std::move(promise)));
+        return future;
     }
 
 private:
     IDatamodelPtr datamodel_;
+    std::vector<std::pair<ITransactionPtr ,std::promise<CommitStatus>> > transactionPtrQueue_;
+    bool stopThread = false;
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: Переменные, которые могут ставиться из разных потоков лучше делать `std::atomic_bool`

File: lib/src/ssyp_db.cpp [hunk 8]
```diff
@@ -70,14 +75,50 @@ class SsypDB : public ISsypDB {
...
+            InitializeCommitThread();
+        }
+        std::lock_guard<std::mutex> lock(mutex);
+        transactionPtrQueue_.push_back(std::make_pair(tx, std::move(promise)));
+        return future;
     }
 
 private:
     IDatamodelPtr datamodel_;
+    std::vector<std::pair<ITransactionPtr ,std::promise<CommitStatus>> > transactionPtrQueue_;
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: У нас везде названия переменных в snake_case

**Review by myxo — CHANGES_REQUESTED** [review]

File: lib/src/ssyp_db.cpp [hunk 9]
```diff
@@ -70,14 +86,51 @@ class SsypDB : public ISsypDB {
...
+
+    void CommitThreadCycle() {
+        while (!stop_thread_.load()) {
+            std::unique_lock<std::mutex> locker(mutex_);
+            while (!thread_notified.load()) queue_check_.wait(locker);
+
+            if (!transaction_ptr_queue_.empty()) {
+                decltype(transaction_ptr_queue_) local_transaction_queue;
+                {
+                    std::lock_guard lock(swap_mutex_);
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: Ну тут ты немножко перестарался =)
Мьютекс делают не на операции, а на ресурс, который они охраняют. В данном случае ты охраняешь transaction_ptr_queue_ и все операции над ним должны быть под мьютексом. А swap_mutex_ - он лишний

File: lib/src/ssyp_db.cpp [hunk 10]
```diff
@@ -70,14 +86,51 @@ class SsypDB : public ISsypDB {
...
+    std::thread commit_thread_;
+    std::mutex mutex_;
+    std::mutex swap_mutex_;
+    std::atomic_bool thread_notified;
+    std::condition_variable queue_check_;
+
+    void CommitThreadCycle() {
+        while (!stop_thread_.load()) {
+            std::unique_lock<std::mutex> locker(mutex_);
+            while (!thread_notified.load()) queue_check_.wait(locker);
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: Да, это делается так. Единственное что, в queue_check_ можно вторым параметром передать лямбду, которая возвращает bool, говорящий о том нужно ли просыпаться или нет. Такой вариант немного получше
https://en.cppreference.com/w/cpp/thread/condition_variable

И ещё, тебе не нужен thread_notified. У тебя уже есть событие по которому нужно выполнять работу - если transaction_ptr_queue_ не пустой

File: lib/src/ssyp_db.cpp [hunk 11]
```diff
@@ -70,14 +86,51 @@ class SsypDB : public ISsypDB {
...
+
+            if (!transaction_ptr_queue_.empty()) {
+                decltype(transaction_ptr_queue_) local_transaction_queue;
+                {
+                    std::lock_guard lock(swap_mutex_);
+                    std::swap(local_transaction_queue, transaction_ptr_queue_);
+                }
+
+                for (auto& [tx, promise] : local_transaction_queue) {
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: А отсюда можно уже снимать locker, так как очередь уже локальная

**Review by myxo — CHANGES_REQUESTED** [review]

File: lib/src/ssyp_db.cpp [hunk 12]
```diff
@@ -10,7 +17,15 @@ class Transaction : public ITransaction {
 class SsypDB : public ISsypDB {
 public:
     SsypDB(DbSettings settings)
-        : datamodel_(CreateDatamodel(CreateStorage(settings), settings)) {}
+        : datamodel_(CreateDatamodel(CreateStorage(settings), settings)) {
+        commit_thread_ = std::thread(&SsypDB::CommitThreadCycle, this);
+        stop_thread_ = false;
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: Ну можно и тут, но лучше прям где объявляешь переменную ниже

File: lib/src/ssyp_db.cpp [hunk 13]
```diff
@@ -70,14 +85,48 @@ class SsypDB : public ISsypDB {
...
+    std::vector<std::pair<ITransactionPtr, std::promise<CommitStatus>>>
+        transaction_ptr_queue_;
+    std::atomic_bool stop_thread_;
+    std::thread commit_thread_;
+    std::mutex mutex_;
+    std::condition_variable queue_check_;
+
+    void CommitThreadCycle() {
+        while (!stop_thread_.load()) {
+            if (!transaction_ptr_queue_.empty()) {
```
Comment (myxo) [inline lib/src/ssyp_db.cpp]: Ещё один момент, обращение к transaction_ptr_queue_ должно быть под мьютексом. В принципе можно это условие вообще убрать и поставить сюда `queue_check_.wait(...)`, он перед тем как начать ждать проверяет условие

File: lib/src/ssyp_db.cpp [hunk 14]
```diff
@@ -70,14 +85,48 @@ class SsypDB : public ISsypDB {
...
     IDatamodelPtr datamodel_;
+    std::vector<std::pair<ITransactionPtr, std::promise<CommitStatus>>>
+        transaction_ptr_queue_;
+    std::atomic_bool stop_thread_;
+    std::thread commit_thread_;
+    std::mutex mutex_;
+    std::condition_variable queue_check_;
+
+    void CommitThreadCycle() {
+        while (!stop_thread_.load()) {
```
Comment (myxo) [inline lib/src/ssyp_db.cpp:107]: Ещё после цикла нужно проверить, если в очереди ещё что-то лежит, но у всех promise нужно сделать set_value Error

**Review by myxo — CHANGES_REQUESTED** [review]

File: lib/src/ssyp_db.cpp [hunk 15]
```diff
@@ -70,14 +84,49 @@ class SsypDB : public ISsypDB {
...
+            for (auto& [tx, promise] : local_transaction_queue) {
+                auto status =
+                    datamodel_->Commit(((Transaction*)(tx.get()))->ops)
+                        ? CommitStatus::Success
+                        : CommitStatus::Error;
+                promise.set_value(status);
+            }
+        }
+        std::lock_guard lock(mutex_);
+        if (!transaction_ptr_queue_.empty()) {
```
Comment (myxo) [inline lib/src/ssyp_db.cpp:124]: Ну этот if можно было и не делать. Если очередь пустая, то и в цикл не зайдет

**Review by myxo — APPROVED** [review]
